# Introduction

https://www.kaggle.com/datasets/amananandrai/ag-news-classification-dataset

#### NEWS LETTER  website wanted to have a model that predicts the category of news by the content of the text news. It is an important task for them because they want to predict which text is in which class so they can have an overview of the news. there are some sources for collecting data such as social media websites and even new files but rarely do they have lable.  some even have but still, need time and money to collect them so I use an open-source dataset that was relevant, and also the size of the data set is big enough to find the best machine learning model. I formulate this problem as an NLP text classification pipeline and since the data set has lable it is supervised machine learning and because the target labels are from different classes then it is a multi clasiffication task. In this pipeline, I do data exploration, preprocessing, feature engineering,on the data set ,and then I try to find the best model with the best hyperparameters. Finally, I would like to calculate the necessary metrics.

# Importin and Loding the data 

## Importing Library

In [35]:
import pandas as pd
import requests
import bs4
import textacy
import spacy
nlp = spacy.load("en_core_web_sm")
import pandas as pd
pd.set_option("max_colwidth", None)
import sklearn.feature_extraction.text
import sklearn.metrics.pairwise
import sklearn.model_selection

import nltk. corpus
import gensim.models
import sklearn.metrics
import sklearn.model_selection
import sklearn.ensemble
import sklearn.naive_bayes
import sklearn.model_selection
import sklearn.feature_extraction.text
import nltk
import gensim
import spacy
import textacy
nlp = spacy.load("en_core_web_sm")
import sklearn.ensemble
import imblearn.over_sampling


## Loding text 

In [36]:
df = pd.read_csv("../../datasets/Newsdataset.csv")
df.head()

Unnamed: 0  \
0           0   
1           1   
2           2   
3           3   
4           4   

                                                            news_headline  \
0        50-year-old problem of biology solved by Artificial Intelligence   
1     Microsoft Teams to stop working on Internet Explorer 11 from Nov 30   
2  Hope US won't erect barriers to cooperation: China on blacklist report   
3      Global smartphone sales in Q3 falls 5.7% to 36.6 cr units: Gartner   
4         EU hoping Biden will clarify US position on digital tax: Report   

                                                                                                                                                                                                                                                                                                                                                                             news_article  \
0             DeepMind's AI system 'AlphaFold' has been recognised as a solution to "protein folding", a grand challenge in biology for over 50 years. DeepMind showed it can predict how proteins fold into 3D shapes, a complex process that is fundamental to understanding the biological machinery of life. AlphaFold can predict the shape of proteins within the width of an atom.   
1                             Microsoft Teams will stop working on Internet Explorer 11 from today, November 30. Users will now need to download the Teams app or browse it on Microsoft Edge. Besides this, Microsoft 365 apps and services will stop supporting Internet Explorer 11 by August 2021. Microsoft will also pull support for Microsoft Edge Legacy desktop app on March 9.   
2  China, in response to reports of US adding Chinese chipmaker SMIC and national offshore oil producer CNOOC to a defence blacklist, said that it hoped US will not erect barriers and obstacles to cooperation. Reuters reported that President Donald Trump's administration is poised to add SMIC and CNOOC to a list of companies allegedly owned or controlled by Chinese military.   
3                                                        The global smartphone sales in the third quarter of 2020 fell by 5.7% year-on-year to 36.6 crore units, according to Gartner. Overall global mobile phone sales to end users totalled 40.1 crore units, a decline of 8.7% year-on-year. Samsung held the top position with a market share of 22%, followed by Huawei with 14.1%.   
4                                                         The European Union (EU) is hoping that US President-elect Joe Biden's incoming administration will clarify US position on digital taxation within two months of taking office, a French Finance Ministry source told Reuters. The bloc-wide tax would apply on digital services offered by companies such as Google and Amazon.   

  news_category  
0    technology  
1    technology  
2    technology  
3    technology  
4    technology

# Exploration, Preprocessing and Feature engineering

In [37]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13166 entries, 0 to 13165
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     13166 non-null  int64 
 1   news_headline  13166 non-null  object
 2   news_article   13166 non-null  object
 3   news_category  13166 non-null  object
dtypes: int64(1), object(3)
memory usage: 411.6+ KB


#### The data set has 13166 rows and 4 columns 3 columns are objects and one intigers.

#### the classes of the news is :

In [38]:
df["news_category"].unique() 

array(['technology', 'sports', 'world', 'politics', 'entertainment',
       'automobile', 'science'], dtype=object)

In [39]:
df.isnull().sum()

Unnamed: 0       0
news_headline    0
news_article     0
news_category    0
dtype: int64

#### Just as info and this isnull method shows there is not any explicit missing values.

### Fixed the news article as variable and news category as target label.

In [40]:
x=df["news_article"]
y=df["news_category"]

In [41]:
x_train, x_test, y_train, y_test = sklearn.model_selection.train_test_split(x, y,test_size=.3,train_size=.7)

#### Split the data as a test and train set and the proportion of test set is 30%and train set is 70% so we have more data to train and the model has more chance to train.

In [42]:
y_train.value_counts()

world            1555
entertainment    1544
sports           1423
technology       1378
politics         1205
science          1090
automobile       1021
Name: news_category, dtype: int64

#### Controlling the balance of data set technology news, politics, science and automobile is not balanced. so one way to deal with this issue is over sampling technic.

#### I use the over sampeling technic to balance the tarin data in this way our model train better and put the random state is a control the randomization of the algorithm and I put it zero.

In [43]:
label_y = sklearn.preprocessing.LabelEncoder()
y_test = label_y.fit_transform(y_test)
y_train = label_y.fit_transform(y_train)

#### Since the category of the news was objected it was necessary to encode them.

In [44]:
vectorizer = sklearn.feature_extraction.text.TfidfVectorizer(min_df=20)
vectorizer.fit(x_train)
x_train = vectorizer.transform(x_train).toarray()
x_test = vectorizer. transform(x_test).toarray()

#### TFidf Is a word vectorazation. It means, it converts the text into vectors and codes them so in this way text be able to be readable for computers. It can notice the importance of a given word. It can quantify the importance of a score for each word measured often terms, and it is useful for classification. itself contains word tokenizer lowercase.

In [49]:
balance = imblearn.over_sampling.RandomOverSampler(random_state=0)
x_train, y_train = balance.fit_resample(x_train,y_train)

#### I use the over sampeling technic to balance the tarin data in this way our model train better and put the random state is a control the randomization of the algorithm and I put it zero.

### modeling traditional model 

In [64]:
parameters_grid = {"n_estimators":[200,250,300],
                    "max_depth":[60,70,80],
                    "min_samples_split":[2,3,5]
}
model_ETC = sklearn.model_selection.GridSearchCV(sklearn.ensemble.ExtraTreesClassifier(),
                                               parameters_grid, scoring="accuracy", cv=5, n_jobs=1)
model_ETC.fit(x_train, y_train)
print("Accuracy of best  classfier = {:.2f}".format(model_ETC.best_score_))
print("Best found hyperparameters of classifier = {}".format(model_ETC.best_params_))

Accuracy of best  classfier = 0.93
Best found hyperparameters of classifier = {'max_depth': 80, 'min_samples_split': 2, 'n_estimators': 250}


In [56]:
y_predicted = model_ETC.predict(x_test)

ac = sklearn.metrics.accuracy_score(y_test,y_predicted)
ac

0.9377215189873418

### Advance model 

In [60]:
label = sklearn.preprocessing.LabelEncoder()
df["news_category"] = label.fit_transform(df["news_category"])

x_new_train,x_new_test,y_new_train,y_new_test = sklearn.model_selection.train_test_split(df["news_article"],
                                                                
                                                df["news_category"],test_size=.4,train_size=.6)



In [61]:
df_train = pd.DataFrame({"text":x_new_train, "labels":y_new_train})
df_test  = pd.DataFrame({"text":x_new_test, "labels":y_new_test})

In [62]:
import simpletransformers.classification

model__simpleTransFormers = simpletransformers.classification.ClassificationModel("roberta","roberta-base",
                                                                                   
       use_cuda = False, num_labels = y.nunique(),
           
         args  = {"overwrte_out_put":True,
              "save_model_every_epoch":False,
              "num_train_epochs":1
}                                                                         
                                                                                  
)

model__simpleTransFormers.train_model(x_train,y_train)

Some weights of the model checkpoint at roberta-base were not used when initializing RobertaForSequenceClassification: ['lm_head.bias', 'lm_head.dense.bias', 'lm_head.dense.weight', 'lm_head.layer_norm.weight', 'roberta.pooler.dense.weight', 'lm_head.layer_norm.bias', 'lm_head.decoder.weight', 'roberta.pooler.dense.bias']
- This IS expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing RobertaForSequenceClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of RobertaForSequenceClassification were not initialized from the model checkpoint at roberta-base and are newly initialized: ['classifier.out_proj.weight', 'classi

AttributeError: 'numpy.ndarray' object has no attribute 'columns'

In [ ]:
model = gensim.models.KeyedVectors.load_word2vec_format("/tmp/GoogleNews-vectors-negative300.bin.gz", binary=True)

Stack Overflow. (n.d.). regex - Faster way to remove stop words in Python. [online] Available at: https://stackoverflow.com/questions/19560498/faster-way-to-remove-stop-words-in-python [Accessed 21 Jun. 2022].